### ISSUE TITLE
Thumbnail Requests fail with 404 errors

### ISSUE THREAD

**Issue by cliffm** [briis/unifiprotect#86]
I have an automation in Node Red that uses the Binary Sensor motion sensor from unifi_protect to send me a telegram message with a thumbnail from Unifi. Every time it triggers, the thumbnail is not ready (likely a delay for Unifi - not sure why). I have handling for the error/failure in Node Red, so it retries, and most of the time it succeeds, but I would prefer to have the retries handled in the component, if it is possible.

I note you are using aiohttp to do the request. There is a aiohttp-retry module, but maybe you can't use it because it's not packaged with Home Assistant. Anyway, I could try to fix it using just aiohttp, if you are open to it.

Example error message, which repeats 3-4 times for every motion detect event:

```2020-06-17 17:24:55 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.140097609497232] Thumbnail Request failed: 404 - Reason: Not Found
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 130, in handle_call_service
    connection.context(msg),
  File "/usr/src/homeassistant/homeassistant/core.py", line 1259, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1294, in _execute_service
    await handler.func(service_call)
  File "/config/custom_components/unifiprotect/__init__.py", line 160, in async_save_thumbnail
    await async_handle_save_thumbnail_service(hass, call)
  File "/config/custom_components/unifiprotect/__init__.py", line 258, in async_handle_save_thumbnail_service
    await _write_thumbnail(camera_id, filename, image_width)
  File "/config/custom_components/unifiprotect/__init__.py", line 244, in _write_thumbnail
    camera_id, image_width
  File "/config/custom_components/unifiprotect/unifi_protect_server.py", line 362, in get_thumbnail
    f"Thumbnail Request failed: {response.status} - Reason: {response.reason}"
custom_components.unifiprotect.unifi_protect_server.NvrError: Thumbnail Request failed: 404 - Reason: Not Found```

**Comment by cliffm** [comment]
I noticed that I was slightly out of date. I was still using the custom_component version, and not the integration. I have now updated, and I get the same problem, albeit in a file with a different name:

```2020-06-17 18:49:09 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection.140318368542352] Thumbnail Request failed: 404 - Reason: Not Found
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 130, in handle_call_service
    connection.context(msg),
  File "/usr/src/homeassistant/homeassistant/core.py", line 1259, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1294, in _execute_service
    await handler.func(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 514, in handle_service
    required_features,
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 454, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 597, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 485, in _handle_entity_call
    await result
  File "/config/custom_components/unifiprotect/camera.py", line 139, in async_save_thumbnail
    image = await self.upv_object.get_thumbnail(self._camera_id, image_width)
  File "/usr/local/lib/python3.7/site-packages/pyunifiprotect/unifi_protect_server.py", line 446, in get_thumbnail
    f"Thumbnail Request failed: {response.status} - Reason: {response.reason}"
pyunifiprotect.unifi_protect_server.NvrError: Thumbnail Request failed: 404 - Reason: Not Found```

**Comment by briis** [comment]
Hi @cliffm 
The thumbnail is created once the Motion Event ends, so you might see the delays you are experiencing. I am open to any suggestions, on how to improve, so please have a look at the [low level](https://github.com/briis/pyunifiprotect) module, and make a PR with your suggestions. 

As this is retrieved from PyPi.org, I can put *aiohttp-retry* in as a requirement and it will be downloaded on activation of the integration. So that should not be a problem.

**Comment by briis** [comment]
Are you still having this issue? If so please let me know if the errors are still the same, else please close this issue. Thx.

**Comment by TazUk** [comment]
Not OP, however using YAML automations I do see a 404 failure from time to time when requesting a thumbnail despite having a delay on the automation:
**Automation**
```
- id: unifiprotect_driveway_motion_notification
  alias: "Driveway Motion Notification"
  mode: queued
  trigger:
    platform: state
    entity_id: binary_sensor.motion_driveway
    from: "on"
    to: "off"
  condition: []
  action:
    - delay: "00:00:05"
    - condition: state
      entity_id: binary_sensor.motion_driveway
      state: "off"
    - service: unifiprotect.save_thumbnail_image
      data_template:
        entity_id: camera.driveway
        filename: "{{ states('sensor.snapshot_local') }}driveway_motion.jpg"
        image_width: 2160

```

**Logged error**
```
Logger: homeassistant.components.automation
Source: custom_components/unifiprotect/camera.py:144
Integration: Automation (documentation, issues)
First occurred: 5 August 2020, 15:02:33 (2 occurrences)
Last logged: 5 August 2020, 15:02:33

Driveway Motion Notification: Error executing script. Unexpected error for call_service at pos 3: Thumbnail Request failed: 404 - Reason: Not Found
While executing automation automation.driveway_motion_notification
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/script.py", line 194, in _async_step
    self, f"_async_{cv.determine_script_action(self._action)}_step"
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/script.py", line 382, in _async_call_service_step
    await service_task
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/core.py", line 1281, in async_call
    task.result()
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/core.py", line 1316, in _execute_service
    await handler.func(service_call)
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity_platform.py", line 527, in handle_service
    required_features,
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/service.py", line 454, in entity_service_call
    future.result()  # pop exception if have
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/entity.py", line 595, in async_request_call
    await coro
  File "/srv/homeassistant/lib/python3.7/site-packages/homeassistant/helpers/service.py", line 485, in _handle_entity_call
    await result
  File "/home/homeassistant/.homeassistant/custom_components/unifiprotect/camera.py", line 144, in async_save_thumbnail
    image = await self.upv_object.get_thumbnail(self._camera_id, image_width)
  File "/srv/homeassistant/lib/python3.7/site-packages/pyunifiprotect/unifi_protect_server.py", line 474, in get_thumbnail
    f"Thumbnail Request failed: {response.status} - Reason: {response.reason}"
pyunifiprotect.unifi_protect_server.NvrError: Thumbnail Request failed: 404 - Reason: Not Found
```

**Comment by briis** [comment]
This an old issue, which I will close now. But for the size of Image you request in your automation, I would recommend to use the `camera.snapshot` service instead of the Thumbnail. This will most likely be removed in the near future anyway.

**Comment by TazUk** [comment]
Hi @briis 

I use the thumbnail service specifically because camera snapshot has a delay, e.g. testing today by triggering the camera snapshot at 11:30:00 returns a snapshot with a camera timestamp at 11:29:50.

Using the thumbnail service means I can rely on the Protect thumbnail, which usually has the motion cause included.

If I have to rely on camera.snapshot with the current delay the image will not include the motion trigger event making the notification useless.

**Comment by briis** [comment]
Fair enough. I was convinced nobody used the service. But you do, so I will leave it in there. Are you still having the issues?

**Comment by TazUk** [comment]
Thank you. Obviously if the HA level issues with camera delay are ever resolved I'm happy to move across to the native service.

The issue still appears occasionally where there's a lot of motion events in a short space of time - approximately once every few days. I can live with it if it's impractical to fix (especially if there are few users of the service).

**Comment by briis** [comment]
Are you on a CloudKey or UDMP/UNVR? 
The next release (In beta now) is moving all the event handling to Websockets, so we go from a Pull to a Push model. That takes a lot of strain off the system, and will most likely fix this issue.
Unfortunately, the Cloudkey does not support Websockets, so it will have an impact on UnfiOS servers only.
If you are on one of these, please try out the Beta that is available in HACS.

**Comment by TazUk** [comment]
Sadly I went with a Cloud Key before the udmp was announced.

**Comment by briis** [comment]
I actually did this too. But as I moved to a new house and wanted to get more control of my network, I decided to take the plunge, and bought a UDMP. This has been more unstable than the CloudKey, but with the latest firmware updates, its running well now - and I now can activate Intruder Protection without slowing down my network.

I heard rumors that Ubiquiti will move the CloudKey to UnifiOS, and then we have no more problems (With this). Would make my life easier, not having to support all these different systems, with different versions.
